src/server/pfs/server/driver_fsck.go

```diff
@@ -318,29 +318,46 @@ func fsckCommits(commitInfos map[string]*pfs.CommitInfo, onError func(error) err
 	return nil
 }
 
-func (d *driver) listReferencedCommits(tx *pachsql.Tx) (map[string]*pfs.Commit, error) {
-	cs := make(map[string]*pfs.Commit)
-	var ids []string
-	if err := tx.Select(&ids, `SELECT commit_id from  pfs.commit_totals`); err != nil {
-		return nil, errors.Wrap(err, "select commit ids from pfs.commit_totals")
+func (d *driver) listReferencedCommits(ctx context.Context, tx *pachsql.Tx) (map[pfsdb.CommitID]*pfsdb.CommitWithID, error) {
+	referencedCommits := make(map[pfsdb.CommitID]*pfsdb.CommitWithID)
+	var ids []pfsdb.CommitID
+	fillMap := func() error {
+		for _, id := range ids {
+			commitID := pfsdb.CommitID(id)
+			if _, ok := referencedCommits[id]; ok {
+				continue
+			}
+			commit, err := pfsdb.GetCommit(ctx, tx, commitID)
+			if err != nil {
+				return errors.Wrap(err, "list referenced commits")
+			}
+			referencedCommits[id] = &pfsdb.CommitWithID{
+				ID:         commitID,
+				CommitInfo: commit,
+			}
+		}
+		return nil
+	}
+	if err := tx.Select(&ids, `SELECT int_id FROM pfs.commits WHERE total_fileset_id IS NOT NULL`); err != nil {
+		return nil, errors.Wrap(err, "list referenced commits: commit totals")
 	}
-	for _, id := range ids {
-		cs[id] = pfsdb.ParseCommit(id)
+	if err := fillMap(); err != nil {
+		return nil, err
 	}
-	ids = make([]string, 0)
-	if err := tx.Select(&ids, `SELECT commit_id from  pfs.commit_diffs`); err != nil {
-		return nil, errors.Wrap(err, "select commit ids from pfs.commit_diffs")
+	ids = make([]pfsdb.CommitID, 0)
+	if err := tx.Select(&ids, `SELECT c.int_id FROM pfs.commits c JOIN pfs.commit_diffs cd ON cd.commit_id = c.commit_id`); err != nil {
+		return nil, errors.Wrap(err, "list referenced commits: commit diffs")
 	}
-	for _, id := range ids {
-		cs[id] = pfsdb.ParseCommit(id)
+	if err := fillMap(); err != nil {
+		return nil, err
 	}
-	return cs, nil
+	return referencedCommits, nil
 }
 
-func fsckDanglingCommits(ris map[string]*pfs.RepoInfo, cs map[string]*pfs.Commit) []*pfs.Commit {
-	var danglingCommits []*pfs.Commit
+func fsckDanglingCommits(ris map[string]*pfs.RepoInfo, cs map[pfsdb.CommitID]*pfsdb.CommitWithID) []*pfsdb.CommitWithID {
```

Comment:
Is this functionality even needed after the postgres migration? I think we have a foreign key constraint now that guarantees that this cannot occur?

Reply:
Yeah that's right. We can probably remove this check then.